src/main/java/org/opensearch/flowframework/workflow/ReIndexStep.java

```diff
@@ -0,0 +1,175 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.workflow;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.action.support.PlainActionFuture;
+import org.opensearch.client.Client;
+import org.opensearch.common.Booleans;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.flowframework.exception.FlowFrameworkException;
+import org.opensearch.flowframework.exception.WorkflowStepException;
+import org.opensearch.flowframework.indices.FlowFrameworkIndicesHandler;
+import org.opensearch.flowframework.util.ParseUtils;
+import org.opensearch.index.reindex.BulkByScrollResponse;
+import org.opensearch.index.reindex.ReindexAction;
+import org.opensearch.index.reindex.ReindexRequest;
+
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.flowframework.common.CommonValue.DESTINATION_INDEX;
+import static org.opensearch.flowframework.common.CommonValue.MAX_DOCS;
+import static org.opensearch.flowframework.common.CommonValue.REFRESH;
+import static org.opensearch.flowframework.common.CommonValue.REQUESTS_PER_SECOND;
+import static org.opensearch.flowframework.common.CommonValue.REQUIRE_ALIAS;
+import static org.opensearch.flowframework.common.CommonValue.RE_INDEX_FIELD;
+import static org.opensearch.flowframework.common.CommonValue.SLICES;
+import static org.opensearch.flowframework.common.CommonValue.SOURCE_INDEX;
+
+/**
+ * Step to reindex
+ */
+public class ReIndexStep implements WorkflowStep {
+
+    private static final Logger logger = LogManager.getLogger(ReIndexStep.class);
+    private final Client client;
+    private final FlowFrameworkIndicesHandler flowFrameworkIndicesHandler;
+
+    /** The name of this step, used as a key in the template and the {@link WorkflowStepFactory} */
+    public static final String NAME = "reindex";
+
+    /**
+     * Instantiate this class
+     *
+     * @param client Client to create an index
+     * @param flowFrameworkIndicesHandler FlowFrameworkIndicesHandler class to update system indices
+     */
+    public ReIndexStep(Client client, FlowFrameworkIndicesHandler flowFrameworkIndicesHandler) {
+        this.client = client;
+        this.flowFrameworkIndicesHandler = flowFrameworkIndicesHandler;
+    }
+
+    @Override
+    public PlainActionFuture<WorkflowData> execute(
+        String currentNodeId,
+        WorkflowData currentNodeInputs,
+        Map<String, WorkflowData> outputs,
+        Map<String, String> previousNodeInputs,
+        Map<String, String> params
+    ) {
+
+        PlainActionFuture<WorkflowData> reIndexFuture = PlainActionFuture.newFuture();
+
+        Set<String> requiredKeys = Set.of(SOURCE_INDEX, DESTINATION_INDEX);
```

Comment:
Should this be `SOURCE_INDICES`?  The `ReindexRequest` supports this and you use `sourceIndices` as the variable to store this value.